### DISCUSSION TITLE
Controlling SGTK config dependencies with cpenv

### DISCUSSION THREAD

**Discussion by fabiangeisler** [cpenv/cpenv#54] [Q&A]
We have some studio-wide python libraries that we use in our SGTK config, which I would like to control with `cpenv`. Since our config is used by ShotGrid Desktop we a bit to early for the app launch hook and cannot control it with the tk-cpenv GUI. What would be the best approach to control extra python libraries for your SGTK config? 

At the moment we use the `tank_init.py` core hook and append all libs we need to pythons `sys.path`, which is as bad as it is easy and straight forward. I am thinking that it is probably possible to get a hold on `cpenv` that is located in `tk-cpenv` and resolve some custom environment for the config. The `tank_init.py` might be too early for that, but I think that is ok. As long as I can use the python library in the app hooks I should be fine. 
Another idea would be to use `cpenv` to create an environment for SG Desktop. That would be the cleanest code wise but would mean that we need to install `cpenv` on all artists machines. And also create a launch script for SG Desktop. Doable but not super convenient. 

I am open for ideas and hints of any kind. 🙂

**Comment by danbradham** [comment]
Another great topic! We personally package our sg config dependencies directly into our zipped configs and load them as needed in various hooks.

I concur that in an ideal world we could manage these with cpenv too. I'm not entirely sure of what the implications would be in activating modules during tank_init. What happens when you switch projects? I think SG desktop might use a separate process for each loaded config but I could be wrong. If that is the case it might be worth standardizing a tank_init hook in tk-cpenv to load some modules or at least providing an example. If not, I'd fear cross-contamination between configs.

I'll have to think about this a bit more when I'm back. Keep me posted on your progress here!

**Comment by fabiangeisler** [comment]
I gave this some thought and it is super tricky. Avoiding cross-contamination is almost impossible.

The main thing I want to have this for, is the standalone Publisher that you can launch from SG Desktop. Since it is registered as an app to SG Desktop, it is not started via the launchapp/ cpenv hooks. But I am using some custom python libraries in it, that somehow need to be added to the environment. I think I could implement some cpenv caching logic in the `pre_publish` hook, but the main issue with this is cleaning up after myself. Once you imported some modules and added some paths to `sys.path` it is going to be hard to remove them again when you switch from one project to another.
One thinkable way would be to vendorize my custom libraries in a SGTK app or framework. That would guaranty that there is no cross-contamination but it also sounds cumbersome to maintain and kinda misses the point of cpenv. 

Will research further for additional ways to go about it.

**Reply by fabiangeisler** [reply]
New idea: 
How about a tk-cpenv-framework that caches modules inside the framework structure (so inside the python folder of the framework in the sgtk bundle cache) based on some settings that you give it in the config. So something like:

```yaml
tk-cpenv-framework_v1.x.x.settings:
  location: <some descriptor>
  modules:
    some-lib: v1.0
    some-plugin: v2.3
```
This means that you could specify any necessary libraries in the framework and you could access them via the usual stgk functions. Cross determination could be avoided depending on how clever the just-in-time module cache is setup. 

Anyway, I have not thought this through completely, but I think it sounds promising. Maybe you have some additional ideas and thoughts on that? :)
